File: python/n206_ReverseList.py
# https://leetcode-cn.com/problems/reverse-linked-list/
# Definition for singly-linked list.

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        pre = None
        cur = head
        while cur is not None:
            tmp = cur
            cur = cur.next
            tmp.next = pre
            pre = tmp
        return pre
        
    def reverseList_a(self, head: ListNode) -> ListNode:
        if head is None or head.next is None: return head
        p = self.reverseList_a(head.next)
        head.next.next = head
        head.next = None
        return p


if __name__ == "__main__":
    pass

File: cpp/n206_ReverseList.cpp
// https://leetcode-cn.com/problems/reverse-linked-list/
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
using namespace std;


// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == NULL || head -> next == NULL) return head;
        ListNode* p = reverseList(head -> next);
        head -> next -> next = head;
        head -> next = NULL;
        return p;
    }
};

int main(){
    return 0;
}